### PR TITLE
Merge branch-ImproveEventDateAndTime-V1.3

### LINKED ISSUE #97: Abstract DateOfBirth and StartDate
The two fields `DateOfBirth` and `StartDate` of the `Person` and `Event` class respectively, behave quite similarly and thus can be abstracted out under one class which they can implement from.

### PR DESCRIPTION
**Pull request by lfrostbytee**
The two classes "StartDate" and "DateOfBirth" in "Person" and "Event" class are mostly the same.

To ensure good code quality, the two classes can be abstracted as one class called "Date" instead.

Let's abstract StartDate and DateOfBirth into Date

Closes #97

### CODE REVIEW

**Comment by lfrostbytee** [comment]
Accidentally did the work on the wrong branch, would that be any issue?

File: src/main/java/seedu/address/model/date/Date.java [hunk 1]
```diff
@@ -8,12 +8,12 @@
...
 /**
- * Represents a Person's date of birth in the address book.
+ * Represents a date in the application.
  * Guarantees: immutable; is valid as declared in {@link #isValidDate(String)}
  */
-public class DateOfBirth implements Comparable<DateOfBirth> {
+public class Date implements Comparable<Date> {
 
-    public static final String MESSAGE_CONSTRAINTS = "Date of birth must be in format: dd/mm/yyyy";
+    public static final String MESSAGE_CONSTRAINTS = "Dates must be in format: dd/MM/yyyy";
```
Comment (Benjamin-Sim) [inline src/main/java/seedu/address/model/date/Date.java]: Is there a reason why the month `MM` is uppercase but day and year are not?

File: src/main/java/seedu/address/model/date/Date.java [hunk 2]
```diff
@@ -55,39 +54,35 @@ public static boolean isValidDateOfBirth(String test) {
...
 
     @Override
-    public int compareTo(DateOfBirth d) {
+    public int compareTo(Date d) {
         return this.date.compareTo(d.date);
     }
 
     /**
-     * Returns the the String representation of the DateOfBirth in format suitable for storage logging
+     * Returns the String representation of the Date in format suitable for storage logging
```
Comment (Benjamin-Sim) [inline src/main/java/seedu/address/model/date/Date.java]:
```suggestion
     * Returns the String representation of the Date in a format suitable for storage logging
```

**Review by Benjamin-Sim — COMMENTED** [review]
All good 👍, just some small typos to fix

File: src/main/java/seedu/address/model/date/Date.java (hunk 1 above)
Comment (Vinodjayakumar124) [inline src/main/java/seedu/address/model/date/Date.java]: Understand that internally it must be represented as MM when use DateTimeFormatter as mm represents minutes, but since this is a user-facing message, I think it'll be better if we continue to use lowercase so as not confuse them with the technicalities of the backend. What do ya think? 

```suggestion
    public static final String MESSAGE_CONSTRAINTS = "Dates must be in format: dd/mm/yyyy";
```

File: src/main/java/seedu/address/model/date/Date.java [hunk 3]
```diff
@@ -25,27 +25,26 @@ public class DateOfBirth implements Comparable<DateOfBirth> {
...
     private static final DateTimeFormatter outputFormatter = DateTimeFormatter.ofPattern("MMM d, yyyy");
 
-
     public final LocalDate date;
 
     /**
-     * Constructs an {@code DateOfBirth}.
+     * Constructs a {@code Date}.
      *
      * @param date A valid date.
```
Comment (Vinodjayakumar124) [inline src/main/java/seedu/address/model/date/Date.java]:
```suggestion
     * @param date a valid date.
```

File: src/main/java/seedu/address/model/date/Date.java [hunk 4]
```diff
@@ -25,27 +25,26 @@ public class DateOfBirth implements Comparable<DateOfBirth> {
...
         this.date = LocalDate.parse(date, logFormatter);
     }
 
     /**
-     * Returns true if a given string is a valid DOB input.
+     * Returns true if a given string is a valid Date input.
      * @return boolean
      */
-
     //found from https://mkyong.com/java/how-to-check-if-date-is-valid-in-java/
```
Comment (Vinodjayakumar124) [inline src/main/java/seedu/address/model/date/Date.java]: Just a minor modification to follow the code reuse guidelines
```suggestion
    //Solution below adapted from https://mkyong.com/java/how-to-check-if-date-is-valid-in-java/
```

**Review by Vinodjayakumar124 — CHANGES_REQUESTED** [review]
Looks good, only some minor changes needed

File: src/main/java/seedu/address/model/date/Date.java (hunk 1 above)
Comment (lfrostbytee) [reply src/main/java/seedu/address/model/date/Date.java]: Ahh I think I didn't catch this. I think I will change it to all uppercase instead as I was confusing it with the DateTimeFormatter (where the cases matter).

Comment (lfrostbytee) [reply src/main/java/seedu/address/model/date/Date.java]: Yup I agree! Thanks for pointing it out!

File: src/main/java/seedu/address/model/date/Date.java [hunk 5]
```diff
@@ -59,8 +59,8 @@ public int compareTo(Date d) {
     }
 
     /**
-     * Returns the String representation of the Date in format suitable for storage logging
-     * @return String
+     * Returns the String representation of the Date in a format suitable for storage logging.
+     * @return String the string that will be logged.
```
Comment (Benjamin-Sim) [inline src/main/java/seedu/address/model/date/Date.java]: "will be logged" sounds like this method does the logging. Should rephrase it to

```suggestion
     * @return String the string that is in the correct format for logging.
```

**Review by Benjamin-Sim — COMMENTED** [review]
Just one more change!

**Review by Vinodjayakumar124 — APPROVED** [review]

**Comment by Vinodjayakumar124** [comment]
All's good!
